runtime/testenv/testenv.go

```diff
@@ -53,7 +54,14 @@ var (
 )
 
 func init() {
-	klog.InitFlags(nil)
+	// NOTE: The following enables glog and klog to coexist. If a test depends
+	// on a package that uses glog, klog flag initialization here would result
+	// in redeclaration of some flags. The following creates a new FlagSet and
+	// uses it to initializes klog.
+	// Refer: https://github.com/kubernetes/klog/blob/f8e668dbaa5f6f0e6a5c24ffd7667263840d79ae/examples/coexist_glog/coexist_glog.go
+	klogFlags := flag.NewFlagSet("klog", flag.ExitOnError)
+	klog.InitFlags(klogFlags)
```

Comment:
Do you need to tell klogr about the flags?

Reply:
OK; I think it is worth pointing out in the signpost comment that it's effectively throwing away the flag values, so that future readers will not skip through and assume they are passed. Since we're talking about global variables here, it would be a reasonable assumption given the comment that say, `klog.InitFlags(...)` _assigns_ the flagset to be used by klogr. (Or does it actually do this? There you go, I'm already confused ...)

Reply:
:smile: yes, it is a bit confusing. I spent some more time going through the klog repo and understanding how it's supposed to be used. It looks like `klog.InitFlags()` is independent of klogr. There's a private variable called `logging` that's used for setting the flag values on the given `FlagSet` to `InitFlags()`.
klogr just imports klog and uses the global klog. There's no API to pass a klog instance. If any values are set on the global klog using the global flagset, it'll  be available to klogr. No way to pass a different klog.

The problem in this case is that since the testenv's init gets called at the very beginning, there's nothing that can be done in testenv package from preventing someone trying to register another flag later. In the case of badger, since they use glog, even they don't have a way to not do this without dropping the glog dependency. The only way around this that I can think of is to avoid filling the global flags with klog init flags for this particular situation.

I'll update the comment to mention that the klog flags are not passed to the klogr that's set as the controller runtime logger. It can be manually passed by creating a custom klog and  klogr, and overwriting the set controller-runtime logger.

Reply:
I've looked at the klog code now too. The flag values aren't thrown away as I speculated, they are used to populate a global struct, as you say. And `klogr` is just an adapter that uses the global. So that's all fine from our point of view, since all we want to do is avoid using the same flagset as glog.